masq_lib/src/percentage.rs

```diff
@@ -1,77 +1,226 @@
 // Copyright (c) 2024, MASQ (https://masq.ai) and/or its affiliates. All rights reserved.
 
-use num::integer::mod_floor;
 use num::CheckedAdd;
 use num::CheckedSub;
 use num::{CheckedDiv, CheckedMul, Integer};
 use std::any::type_name;
 use std::fmt::Debug;
-use std::ops::Div;
-use std::ops::Mul;
+use std::ops::Rem;
+// Designed to store values from 0 to 100 and offer a set of handy methods for PurePercentage
+// operations over a wide variety of integer types. It is also to look after the least significant
+// digit on the resulted number in order to avoid the effect of a loss on precision that genuinely
+// comes with division on integers if a remainder is left over. The percents are always represented
+// by an unsigned integer. On the contrary, the numbers that it is applied on can take on both
+// positive and negative values.
 
-// It's designed for a storage of values from 0 to 100, after which it can be used to compute
-// the corresponding portion of many integer types. It should also take care of the least significant
-// digit in order to diminish the effect of a precision loss genuinly implied by this kind of math
-// operations done on integers.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PurePercentage {
+    degree: u8,
+}
+
+// This is a wider type that allows to specify cumulative percents of more than only 100.
+// The expected use of this would look like requesting percents meaning possibly multiples of 100%,
+// roughly, of a certain base number. Similarly to the PurePercentage type, also signed numbers
+// would be accepted.
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub struct Percentage {
-    per_cent: u8,
+pub struct LoosePercentage {
+    multiples_of_100_percent: u32,
+    degrees_from_remainder: PurePercentage,
+}
+
+pub trait PercentageInteger:
+    TryFrom<i8>
+    + CheckedMul
+    + CheckedAdd
+    + CheckedSub
+    + CheckedDiv
+    + PartialOrd
+    + Rem<Output = Self>
+    + Integer
+    + Debug
+    + Copy
+{
+}
+
+macro_rules! impl_percentage_integer {
+    ($($num_type: ty),+) => {
+        $(impl PercentageInteger for $num_type {})+
+    }
+}
+
+impl_percentage_integer!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128);
+
+impl LoosePercentage {
+    pub fn new(percents: u32) -> Self {
+        let multiples_of_100_percent = percents / 100;
+        let remainder = (percents % 100) as u8;
+        let degrees_from_remainder =
+            PurePercentage::try_from(remainder).expect("should never happen");
+        Self {
+            multiples_of_100_percent,
+            degrees_from_remainder,
+        }
+    }
+
+    // If this overflows you probably want to precede the operation by converting your base number
+    // to a larger integer type
+    pub fn of<N>(&self, num: N) -> Result<N, BaseTypeOverflow>
+    where
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+        N: TryFrom<u32>,
```

Comment:
Similar to how we have `i8` as a boundary in `PercentageInteger`, please add `u32` too.